bciers/apps/reporting/src/app/components/activities/ActivityForm.tsx

```diff
@@ -56,77 +52,45 @@ export default function ActivityForm({
   // ✅ Success state for for the Submit button
   const [isSuccess, setIsSuccess] = useState(false);
   const [formState, setFormState] = useState(activityFormData as any);
-  const [jsonSchema, setJsonSchema] = useState({});
-  const [uiSchema, setUiSchema] = useState({});
-  const [previousActivityId, setPreviousActivityId] = useState<number>();
+  const [jsonSchema, setJsonSchema] = useState(initialJsonSchema);
+  const [selectedSourceTypeIds, setSelectedSourceTypeIds] = useState(
+    initialSelectedSourceTypeIds,
+  );
 
   const { activityId, sourceTypeMap } = activityData;
 
-  // Set useEffect dependency set from checked sourceTypes
-  const selectedSourceTypesArray = Object.values(sourceTypeMap).map(
-    (v) => formState?.[`${v}`] ?? false,
-  );
-  const numberOfSelectedSourceTypes = selectedSourceTypesArray.filter(
-    (x) => x === true,
-  ).length;
-  const dependencyArray = [numberOfSelectedSourceTypes, activityId];
-
   useEffect(() => {
-    let isFetching = true;
+    setJsonSchema(initialJsonSchema);
+    setFormState(activityFormData);
+    setSelectedSourceTypeIds(initialSelectedSourceTypeIds);
+  }, [currentActivity.id]);
+
+  const validator = customizeValidator({});
+
+  const fetchSchemaData = async (sourceTypeIds: string[]) => {
+    let sourceTypeQueryString = "";
+    sourceTypeIds.forEach((id) => {
+      sourceTypeQueryString += `&source_types[]=${id}`;
+    });
+    const schema = await actionHandler(
+      `reporting/build-form-schema?activity=${currentActivity.id}&report_version_id=${reportVersionId}${sourceTypeQueryString}`,
+      "GET",
+      "",
+    );
+    return schema;
+  };
 
-    const fetchSchemaData = async (
-      selectedSourceTypes: string,
-      selectedKeys: number[],
-    ) => {
-      // fetch data from server
-      const schemaData = await actionHandler(
-        `reporting/build-form-schema?activity=${activityId}&report_version_id=${reportVersionId}${selectedSourceTypes}`,
-        "GET",
-        "",
-      );
+  const handleFormChange = async (c: any) => {
+    const selectedSourceTypes = [];
+    // Checks for a change in source type selection & fetches the updated schema if they have changed.
+    for (const [k, v] of Object.entries(sourceTypeMap)) {
+      if (c.formData[`${v}`]) selectedSourceTypes.push(k);
+    }
+    if (selectedSourceTypes.length !== selectedSourceTypeIds.length) {
```

Comment:
The debounce is on the onChange() because it was because it was slow locally when typing (I think because of the size of the rjsf schema). It's not debouncing the call to the API to save the data